lib/CodeGen/CodeGenFunction.h

```diff
@@ -1238,6 +1238,48 @@ class CodeGenFunction : public CodeGenTypeCache {
   const TargetInfo &getTarget() const { return Target; }
   llvm::LLVMContext &getLLVMContext() { return CGM.getLLVMContext(); }
 
+public:
+  enum BoundsCheckKind {
+    BC_None,
+    BC_Deref,
+    BC_ArraySubscript
+  };
+
+private:
+  BoundsExpr* NextBoundsExprForCheck;
+  BoundsCheckKind NextBoundsCheckKind;
+
+public:
+  BoundsExpr* GetAndClearNextBoundsCheckExpr(BoundsCheckKind Kind) {
+    if (!getLangOpts().CheckedC)
+      return nullptr;
+
+    if (NextBoundsCheckKind == BC_None)
+      return nullptr;
+
+    if (NextBoundsCheckKind != Kind)
+      return nullptr;
+
+    BoundsExpr* result = NextBoundsExprForCheck;
+    NextBoundsExprForCheck = nullptr;
+    NextBoundsCheckKind = BC_None;
+
+    return result;
+  }
+
+  void SetNextBoundsCheckExpr(BoundsExpr *Bounds, BoundsCheckKind Kind) {
+    if (!getLangOpts().CheckedC)
```

Comment:
I think you can delete this test that this is only CheckedC.    Given that you have to pass in a BoundsExpr, it is unlikely to be accidentally called.   If you want to, you could this into an assert.